src/data/nat/choose/central.lean

```diff
@@ -98,4 +108,104 @@ lemma four_pow_le_two_mul_self_mul_central_binom : ∀ (n : ℕ) (n_pos : 0 < n)
 calc 4 ^ n ≤ n * central_binom n : (four_pow_lt_mul_central_binom _ le_add_self).le
 ... ≤ 2 * n * central_binom n    : by { rw [mul_assoc], refine le_mul_of_pos_left zero_lt_two }
 
+variables {p n : ℕ}
+
+/--
+A logarithmic upper bound on the multiplicity of a prime in the central binomial coefficient.
+-/
+lemma padic_val_nat_central_binom_le (hp : p.prime) :
+  padic_val_nat p (central_binom n) ≤ log p (2 * n) :=
+begin
+  rw @padic_val_nat_def _ ⟨hp⟩ _ (central_binom_ne_zero n),
+  unfold central_binom,
+  have two_n_sub : 2 * n - n = n, by rw [two_mul n, nat.add_sub_cancel n n],
+  simp only [nat.prime.multiplicity_choose hp (le_mul_of_pos_left zero_lt_two) (lt_add_one _),
+      two_n_sub, ←two_mul, enat.get_coe', finset.filter_congr_decidable],
+  calc _  ≤ (finset.Ico 1 (log p (2 * n) + 1)).card : finset.card_filter_le _ _
+      ... = (log p (2 * n) + 1) - 1                 : nat.card_Ico _ _,
+end
+
+/--
+Sufficiently large primes appear only to multiplicity 0 or 1 in the central binomial coefficient.
+-/
+lemma padic_val_nat_central_binom_of_large_le_one (hp : p.prime) (p_large : 2 * n < p ^ 2) :
+  (padic_val_nat p (central_binom n)) ≤ 1 :=
+begin
+  have log_weak_bound : log p (2 * n) ≤ 2,
+    { calc log p (2 * n) ≤ log p (p ^ 2) : log_le_log_of_le (le_of_lt p_large)
+      ... = 2 : log_pow hp.one_lt 2, },
+
+  have log_bound : log p (2 * n) ≤ 1,
+    { cases le_or_lt (log p (2 * n)) 1 with log_le lt_log,
+      { exact log_le, },
+      { have v : log p (2 * n) = 2 := by linarith,
+        cases le_or_lt p (2 * n) with h h,
+        { exfalso,
+          rw [log_of_one_lt_of_le hp.one_lt h, succ_inj', log_eq_one_iff] at v,
+          have bad : p ^ 2 ≤ 2 * n,
+            { rw pow_two,
+              exact (nat.le_div_iff_mul_le _ _ (prime.pos hp)).1 v.2.2, },
+          exact lt_irrefl _ (lt_of_le_of_lt bad p_large), },
+        { rw log_eq_zero (or.inl h),
+          exact zero_le 1, }, }, },
+
+  exact le_trans (padic_val_nat_central_binom_le hp) log_bound,
+end
+
+/--
+Sufficiently large primes less than `n` do not appear in the factorisation of `central_binom n`.
+-/
+lemma padic_val_nat_central_binom_of_large_eq_zero
+  (hp : p.prime) (n_big : 2 < n) (p_le_n : p ≤ n) (big : 2 * n < 3 * p) :
+  padic_val_nat p (central_binom n) = 0 :=
+begin
+  rw @padic_val_nat_def _ ⟨hp⟩ _ (central_binom_ne_zero n),
+  unfold central_binom,
+  have two_n_sub : 2 * n - n = n, by rw [two_mul n, nat.add_sub_cancel n n],
+  simp only [
+    nat.prime.multiplicity_choose hp (le_mul_of_pos_left zero_lt_two) (lt_add_one _),
+    two_n_sub, ←two_mul, finset.card_eq_zero, enat.get_coe', finset.filter_congr_decidable],
+  clear two_n_sub,
+
+  have three_lt_p : 3 ≤ p,
+  { rcases le_or_lt 3 p with H|H,
+    { exact H, },
```

Comment:
Does linarith need this case distinction? Seems a bit surprising to me.

Reply:
You're quite right, it doesn't!